packages/wonder-blocks-data/util/request-tracking.js

```diff
@@ -0,0 +1,81 @@
+// @flow
+import * as React from "react";
+
+import type {IRequestHandler} from "./types.js";
+
+type TrackerFn = (handler: IRequestHandler<any, any>, options: any) => void;
+
+type HandlerCache = {
+    [type: string]: IRequestHandler<any, any>,
+    ...,
+};
+
+type RequestCache = {
+    [handlerType: string]: {
+        [key: string]: any,
+    },
+    ...,
+};
+
+/**
+ * Used to inject our tracking function into the render framework.
+ *
+ * INTERNAL USE ONLY
+ */
+export const TrackerContext: React.Context<?TrackerFn> = new React.createContext<?TrackerFn>(
+    null,
+);
+
+export class RequestTracker {
+    /**
+     * These are the caches for tracked requests and their handlers.
+     */
+    _trackedHandlers: HandlerCache = {};
+    _trackedRequests: RequestCache = {};
+
+    /**
+     * Track a request.
+     *
+     * This method caches a request and its handler for use during server-side
+     * rendering to allow us to fulfill requests before producing a final render.
+     */
+    trackDataRequest(handler: IRequestHandler<any, any>, options: any): void {
+        const key = handler.getKey(options);
+        const type = handler.type;
+
+        /**
+         * Make sure we have stored the handler for use when fulfilling requests.
+         */
+        if (this._trackedHandlers[type] == null) {
+            this._trackedHandlers[type] = handler;
+            this._trackedRequests[type] = {};
+        }
```

Comment:
The assumption here is that both `_trackedHandlers[type]` and `this._trackedRequests[type]` are either both undefined or both defined at the same time, right?

Reply:
Yes, that's correct.